diff --git a/include/Common/EABase/int128.h b/include/Common/EABase/int128.h
index 3e78ebc..bab08b2 100644
--- a/include/Common/EABase/int128.h
+++ b/include/Common/EABase/int128.h
@@ -286,6 +286,7 @@ EA_CONSTEXPR inline int128_t_base::operator signed char() const        { return
 EA_CONSTEXPR inline int128_t_base::operator unsigned char() const      { return static_cast<unsigned char>(Low()); }
 EA_CONSTEXPR inline int128_t_base::operator unsigned int() const       { return static_cast<unsigned int>(Low()); }
 EA_CONSTEXPR inline int128_t_base::operator unsigned long long() const { return static_cast<unsigned long long>(Low()); }
+EA_CONSTEXPR inline int128_t_base::operator unsigned long() const      { return static_cast<unsigned long>(Low()); }
 EA_CONSTEXPR inline int128_t_base::operator unsigned short() const     { return static_cast<unsigned short>(Low()); }
 EA_CONSTEXPR inline int128_t_base::operator float() const              { return static_cast<float>(Low()); }
 EA_CONSTEXPR inline int128_t_base::operator double() const             { return static_cast<double>(Low()); }
